### PR TITLE
Dev gate applicator

### PR DESCRIPTION
**Pull request by shahidee44**
## Description

- Changed the way `Gate` Objects are structured. Each non-core openqaoa module will need a `GateApplicator` object which contains library specific functions on how the gates are applied to the circuit. The `Gate` Object now acts as an interface.
- Each Backend now creates a `GateApplicator` Object that, together with the `Gate` Object apply the right library gate to the Backend's internal circuit
- The implementation of `GateApplicator` for Qiskit, Braket and Pyquil backends.
- Dynamic generation of constants: `DEVICE_NAME_TO_OBJECT_MAPPER` , `DEVICE_ACCESS_OBJECT_MAPPER`. Based on the presence of the particular module.
- `get_qaoa_backend` and `create_device` updated to dynamically find a module if it has been installed.
- `plugin_finder.py` is added which helps facilitate the function needed to dynamically find modules within the environment.
- Updated `test_gates.py` unittests
- Added `test_gate_applicator_qiskit.py` unittests
- Added `test_gate_applicator_braket.py` unittests
- Added `test_gate_applicator_pyquil.py` unittests
- Added `test_gate_applicator_vectorized.py` unittests
- Edited `setup.py` to include `entry_points` which will be used by plugins to interact with core code.
- Removed `core` code dependency on any plugin

## Checklist

[//]: <> (- [ ] My code follows the style guidelines of this project)

- [x] I have performed a self-review of my code.
- [ ] I have commented my code and used numpy-style docstrings
- [ ] I have made corresponding updates to the documentation.
- [x] My changes generate no new warnings
- [x] I have added/updated tests to make sure bugfix/feature works.
- [x] New and existing unit tests pass locally with my changes.

[//]: <> (- [ ] Any dependent changes have been merged and published in downstream modules)

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Name the new unit-tests that you have added along with this change.

run all tests.

### CODE REVIEW

**Comment by vishal-ph** [comment]
Since, I proposed the design in this PR, it makes sense that either @raulconchello, @Q-lds, be the primary reviewer of this PR. I am adding you both for now. I will be reviewing it too

File: src/openqaoa-core/backends/plugin_finder.py
Comment (Q-lds) [inline src/openqaoa-core/backends/plugin_finder.py:3]: Can we do all this programmatically?

basically, looking with modules of the form `openqaoa_*` and then creating all the variables programmatically?

File: src/openqaoa-braket/backends/gates_braket.py
Comment (Q-lds) [inline src/openqaoa-braket/backends/gates_braket.py:10]: So, I guess a general design choice could be to replace all this with ENUM... it may be worth considering for a future improvement

File: src/openqaoa-braket/backends/gates_braket.py
Comment (Q-lds) [inline src/openqaoa-braket/backends/gates_braket.py:26]: The ENUM strategy (I think) will also remove the need for these checks (they are basically built-in within the ENUM structure)

File: src/openqaoa-core/backends/qaoa_backend.py
Comment (Q-lds) [inline src/openqaoa-core/backends/qaoa_backend.py:21]: I think from here to line ~39 it should all be done programmatically. Using things like `BACKEND_BOOLEAN_LIST[2]` and associating it strictly yo, say `qiskit`, may cause confusion if the order of plugins is changed

File: src/openqaoa-core/backends/plugin_finder.py
Comment (shahidee44) [reply src/openqaoa-core/backends/plugin_finder.py:3]: Yes this is possible. I also think its possible to go with a more robust/verbose solution that uses a dictionary with key, values here.

File: src/openqaoa-braket/backends/gates_braket.py
Comment (shahidee44) [reply src/openqaoa-braket/backends/gates_braket.py:10]: Should be possible. Need to test how we can do this.

File: src/openqaoa-core/backends/qaoa_backend.py
Comment (shahidee44) [reply src/openqaoa-core/backends/qaoa_backend.py:21]: Agreed. I can change this to a more verbose/strict data structure like a dictionary.

File: src/openqaoa-braket/backends/gates_braket.py
Comment (raulconchello) [inline src/openqaoa-braket/backends/gates_braket.py:24]: Wouldn't it be better to have this as a attribute of the class, not the object.
I mean, instead of:
```
class BraketGateApplicator(gates_core.GateApplicator):
    ...
    def __init__(self):
        self.library = 'braket'
```
this:
```
class BraketGateApplicator(gates_core.GateApplicator):
    ...
    library = 'braket'
```

File: src/openqaoa-core/backends/qaoa_backend.py
Comment (Q-lds) [reply src/openqaoa-core/backends/qaoa_backend.py:70]: Yes, I agree!

File: src/openqaoa-core/backends/gates_vectorized.py
Comment (shahidee44) [reply src/openqaoa-core/backends/gates_vectorized.py:22]: I've removed the try and except in the latest commit.

File: src/openqaoa-braket/backends/gates_braket.py
Comment (shahidee44) [reply src/openqaoa-braket/backends/gates_braket.py:24]: Makes sense. I'll make this change.

File: src/openqaoa-core/backends/gates_vectorized.py
Comment (vishal-ph) [reply src/openqaoa-core/backends/gates_vectorized.py:88]: I think it's a great suggestion to dramatically reduce the number of lines! However, it makes the code less readable IMO. 
@shahidee44, I leave this to you if we want to incorporate this change.

Comment (shahidee44) [reply src/openqaoa-core/backends/gates_vectorized.py:88]: Same sentiments for me. I think it the impact on readability is a no-go for me in this case. Let's KIV this comment in the future and see if there's a possible solution that we could come up with in the future to solve this. For now, i'll resolve this comment.

**Comment by vishal-ph** [comment]
NOTE: If not done yet, run `black` formatter before merging

File: src/openqaoa-azure/utilities.py
Comment (shahidee44) [reply src/openqaoa-azure/utilities.py:1]: Done in latest commit!

**Review by vishal-ph — APPROVED** [review]

**Pull request merged**
